src/util/animate.ts

```diff
@@ -1,34 +1,124 @@
-//@ts-nocheck
 import { fabric } from '../../HEADER';
 import { runningAnimations } from './animation_registry';
 import { noop } from '../constants';
+import { WithReturnType } from '../typedefs';
+import { defaultEasing, EasingFunction } from './anim_ease';
 
 /**
- *
- * @typedef {Object} AnimationOptions
- * Animation of a value or list of values.
- * @property {Function} [onChange] Callback; invoked on every value change
- * @property {Function} [onComplete] Callback; invoked when value change is completed
- * @property {number | number[]} [startValue=0] Starting value
- * @property {number | number[]} [endValue=100] Ending value
- * @property {number | number[]} [byValue=100] Value to modify the property by
- * @property {Function} [easing] Easing function
- * @property {number} [duration=500] Duration of change (in ms)
- * @property {Function} [abort] Additional function with logic. If returns true, animation aborts.
- * @property {number} [delay] Delay of animation start (in ms)
- *
- * @typedef {() => void} CancelFunction
- *
- * @typedef {Object} AnimationCurrentState
- * @property {number | number[]} currentValue value in range [`startValue`, `endValue`]
- * @property {number} completionRate value in range [0, 1]
- * @property {number} durationRate value in range [0, 1]
- *
- * @typedef {(AnimationOptions & AnimationCurrentState & { cancel: CancelFunction }} AnimationContext
+ * Callback called every frame
+ * @param t current "time"/ms elapsed. multivalue
+ * @param valueRatio ratio of current value to animation max value. [0, 1]
+ * @param timeRatio ratio of current ms to animation duration. [0, 1]
  */
+export type OnChangeCallback = (
+  t: number | number[],
+  valueRatio: number,
+  timeRatio: number
+) => void;
```

Comment:
What about making this a generic returning `T` defaulting to `unknown`?
Then you don't need `WithReturnType`